source/common/buffer/buffer_impl.h

```diff
@@ -1,16 +1,434 @@
+   * Copy as much of the supplied data as possible to the end of the slice.
+   * @param data start of the data to copy.
+   * @param size number of bytes to copy.
+   * @return number of bytes copied (may be a smaller than size, may even be zero).
+   */
+  uint64_t append(const void* data, uint64_t size) {
+    if (reservation_outstanding_) {
+      return 0;
+    }
+    uint64_t copy_size = std::min(size, reservableSize());
+    uint8_t* dest = base_ + reservable_;
+    reservable_ += copy_size;
+    // NOLINTNEXTLINE(clang-analyzer-core.NullDereference)
+    memcpy(dest, data, copy_size);
+    return copy_size;
+  }
+
+  /**
+   * Copy as much of the supplied data as possible to the front of the slice.
+   * If only part of the data will fit in the slice, the bytes from the _end_ are
+   * copied.
+   * @param data start of the data to copy.
+   * @param size number of bytes to copy.
+   * @return number of bytes copied (may be a smaller than size, may even be zero).
+   */
+  uint64_t prepend(const void* data, uint64_t size) {
+    if (reservation_outstanding_) {
+      return 0;
+    }
+    const uint8_t* src = static_cast<const uint8_t*>(data);
+    uint64_t copy_size;
+    if (dataSize() == 0) {
+      // There is nothing in the slice, so put the data at the very end in case the caller
+      // later tries to prepend anything else in front of it.
+      copy_size = std::min(size, reservableSize());
+      reservable_ = capacity_;
+      data_ = capacity_ - copy_size;
+    } else {
+      if (data_ == 0) {
+        // There is content in the slice, and no space in front of it to write anything.
+        return 0;
+      }
+      // Write into the space in front of the slice's current content.
+      copy_size = std::min(size, data_);
+      data_ -= copy_size;
+    }
+    memcpy(base_ + data_, src + size - copy_size, copy_size);
+    return copy_size;
+  }
+
+protected:
+  Slice(uint64_t data, uint64_t reservable, uint64_t capacity)
+      : data_(data), reservable_(reservable), capacity_(capacity) {}
+
+  /** Start of the slice - subclasses must set this */
+  uint8_t* base_{nullptr};
+
+  /** Offset in bytes from the start of the slice to the start of the Data section */
+  uint64_t data_;
+
+  /** Offset in bytes from the start of the slice to the start of the Reservable section */
+  uint64_t reservable_;
+
+  /** Total number of bytes in the slice */
+  uint64_t capacity_;
+
+  /** Whether reserve() has been called without a corresponding commit(). */
+  bool reservation_outstanding_{false};
+};
+
+using SlicePtr = std::unique_ptr<Slice>;
+
+class OwnedSlice : public Slice {
+public:
+  /**
+   * Create an empty OwnedSlice.
+   * @param capacity number of bytes of space the slice should have.
+   * @return an OwnedSlice with at least the specified capacity.
+   */
+  static SlicePtr create(uint64_t capacity) {
+    uint64_t slice_capacity = sliceSize(capacity);
+    return SlicePtr(new (slice_capacity) OwnedSlice(slice_capacity));
```

Comment:
This is the 2nd option in the "Syntax" section near the top of https://en.cppreference.com/w/cpp/language/new :
`::`_(optional)_ `new` `(`placement_params`)`_(optional)_ `type` initializer _(optional)_ (2)

As far as I know, they don't show parentheses around the "initializer" part because it's just one of the supported initializer forms (with curly-brace style initializers being the other).

(Interestingly, the first option in that section is the same thing except with parentheses wrapped around 
the type name, which I don't think I've seen used in practice yet.)

This section has a little more detail: https://en.cppreference.com/w/cpp/language/new#Placement_new

Now that you mention it, I have concerns about the idiom used in heap_stat_data.cc, as it explicitly uses `malloc` for the allocation but just lets the compiler use the standard `::delete` implementation for the deallocation. I can't find a reference that guarantees `::delete` will use `free` by default. And if anybody ever plugs in an override for the global `::new` and `::delete` that uses a different underlying API than `malloc`/`free`, that code could stop working.

Reply:
re HeapStatData the constructor and destructor are private; it needs to be freed by calling its class free() method which then calls ::free(). You are absolutely right that deleting a malloced pointer is not safe -- and valgrind  checks this -- but this code can't do that from outside the class because of the private ctor/dtor.

Thanks for the explanation about your use of placed new.

Basically the bottom line is that most modern C++ is insulated from this detail when we are trying to hyper-optimize we need to be more explicit with memory and indirection. To me, the malloc/free/private-dtor pattern is a little clearer -- but that's probably just because I'm used to it.

Reply:
Ah, thanks for the pointer; I missed the private destructor in HeapStatData. The advantage I see in the placement new idiom (compared to using `malloc`) is that it works safely with a public destructor, which in turn enables the use of `unique_ptr` and `shared_ptr`.

Reply:
Actually unique_ptr works fine with a custom free(), and is used for that class:

https://github.com/envoyproxy/envoy/blob/90a7a3eb8f477350a7175cc0d98487bbed1c4188/source/common/stats/heap_stat_data.cc#L23